File: app/src/main/java/com/ladsoft/bakingapp/di/component/AppComponent.java
package com.ladsoft.bakingapp.di.component;

import android.app.Application;

import com.ladsoft.bakingapp.application.BakingAppApplication;
import com.ladsoft.bakingapp.di.module.ActivityBuilder;
import com.ladsoft.bakingapp.di.module.ApiModule;
import com.ladsoft.bakingapp.di.module.AppModule;
import com.ladsoft.bakingapp.di.module.ModelModule;
import com.ladsoft.bakingapp.di.module.RepositoryModule;
import com.ladsoft.bakingapp.di.module.ServiceBuilder;
import com.ladsoft.bakingapp.widget.BakingAppAppWidgetProvider;
import com.ladsoft.bakingapp.widget.remoteviewsservice.IngredientsListWidgetService;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.support.AndroidSupportInjectionModule;

@Singleton
@Component(modules = {
                AndroidSupportInjectionModule.class,
                AppModule.class,
                ApiModule.class,
                RepositoryModule.class,
                ModelModule.class,
                ActivityBuilder.class,
                ServiceBuilder.class
        })
public interface AppComponent {

    void inject(BakingAppApplication application);

    @Component.Builder
    interface Builder {
        @BindsInstance Builder application(Application application);
        AppComponent build();
    }

    void inject(IngredientsListWidgetService.ListRemoteViewsFactory viewFactory);
    void inject(BakingAppAppWidgetProvider provider);
}

File: app/src/main/java/com/ladsoft/bakingapp/di/component/AppComponent.kt
package com.ladsoft.bakingapp.di.component

import android.app.Application
import com.ladsoft.bakingapp.application.BakingAppApplication
import com.ladsoft.bakingapp.di.module.ActivityBuilder
import com.ladsoft.bakingapp.di.module.AppModule
import com.ladsoft.bakingapp.di.module.ModelModule
import com.ladsoft.bakingapp.di.module.RepositoryModule
import com.ladsoft.bakingapp.service.IngredientUpdateService
import com.ladsoft.bakingapp.widget.BakingAppAppWidgetProvider
import com.ladsoft.bakingapp.widget.remoteviewsservice.IngredientsListWidgetService
import dagger.BindsInstance
import dagger.Component
import dagger.android.support.AndroidSupportInjectionModule
import javax.inject.Singleton

@Singleton
@Component(modules = [
//                AndroidInjectionModule::class,
                AndroidSupportInjectionModule::class,
                AppModule::class,
                RepositoryModule::class,
                ModelModule::class,
                ActivityBuilder::class
            ])
interface AppComponent {

    fun inject(application: BakingAppApplication)

    @Component.Builder
    interface Builder {
        @BindsInstance fun application(application: Application): Builder
        fun build(): AppComponent
    }

//    fun inject(recipeState: RecipeState)
    fun inject(viewFactory: IngredientsListWidgetService.ListRemoteViewsFactory)
    fun inject(provider: BakingAppAppWidgetProvider)
    fun inject(service: IngredientUpdateService)

//    object Initializer {
//        fun initialize(application: BakingAppApplication): AppComponent = DaggerAppComponent
//                .builder()
//                .application(application)
//                .build()
//                .inject(application)
//        fun initialize(application: BakingAppApplication): AppComponent = DaggerAppComponent
//                .builder()
//                .appModule(AppModule(application))
//                .build()
//    }
}